core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java

```diff
@@ -0,0 +1,287 @@
+
+            latch.countDown();
+
+            if (randomBoolean()) {
+                throw new ElasticsearchException("this shouldn't cause the test to fail!");
+            }
+        };
+
+        Cancellable cancellable = threadPool.scheduleWithFixedDelay(countingRunnable, TimeValue.timeValueMillis(100L), Names.GENERIC);
+        assertNotNull(cancellable);
+
+        // wait for the number of successful count down operations
+        latch.await();
+
+        // cancel
+        cancellable.cancel();
+        assertTrue(cancellable.isCancelled());
+    }
+
+    public void testCancellingRunnable() throws Exception {
+        final boolean shouldThrow = randomBoolean();
+        final AtomicInteger counter = new AtomicInteger(scaledRandomIntBetween(2, 16));
+        final CountDownLatch doneLatch = new CountDownLatch(1);
+        final AtomicReference<Cancellable> cancellableRef = new AtomicReference<>();
+        final AtomicBoolean runAfterDone = new AtomicBoolean(false);
+        final Runnable countingRunnable = () -> {
+
+            if (doneLatch.getCount() == 0) {
+                runAfterDone.set(true);
+                fail("this runnable was cancelled by itself so it should never run again!");
+            }
+
+            final Cancellable cancellable = cancellableRef.get();
+            if (cancellable == null) {
+                // wait for the cancellable to be present before we really start so we can accurately know we cancelled
+                return;
+            }
+
+            // rarely throw an exception before counting down
+            if (shouldThrow && rarely()) {
+                throw new RuntimeException("throw before count down");
+            }
+
+            final int count = counter.decrementAndGet();
+
+            if (shouldThrow && randomBoolean()) {
+                throw new ElasticsearchException("exception thrown after decrementing counter");
+            }
+
+            // see if we have counted down to zero or below yet. the exception throwing could make us count below zero
+            if (count <= 0) {
+                cancellable.cancel();
+                doneLatch.countDown();
+            }
+
+            // rarely throw an exception after execution
+            if (shouldThrow && rarely()) {
+                throw new RuntimeException("throw at end");
+            }
+        };
+        Cancellable cancellable = threadPool.scheduleWithFixedDelay(countingRunnable, TimeValue.timeValueMillis(50), Names.GENERIC);
+        cancellableRef.set(cancellable);
+        // wait for the runnable to finish
+        doneLatch.await();
+
+        // the runnable should have cancelled itself
+        assertTrue(cancellable.isCancelled());
+        assertFalse(runAfterDone.get());
+
+        // we rely on concurrency to check for the case where a run after cancel happens, but we can rarely do some waiting
+        if (rarely()) {
+            assertFalse(awaitBusy(runAfterDone::get, 2L, TimeUnit.SECONDS));
```

Comment:
this is hard to test since the runnable could run after we check it being false. This rarely awaits to make sure it isn't run. If you don't think it is necessary, I can remove it.